src/cmd/tools/dtest/docker/harness/a_coordinator.go

```diff
@@ -0,0 +1,296 @@
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package harness
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/m3db/m3/src/query/generated/proto/admin"
+	dockertest "github.com/ory/dockertest"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+const (
+	coordinatorDockerfile = "./m3coordinator.Dockerfile"
+)
+
+const (
+	defaultCoordinatorSource     = "coordinator"
+	defaultCoordinatorName       = "coord01"
+	defaultCoordinatorDockerfile = "./m3coordinator.Dockerfile"
+)
+
+func zapMethod(s string) zapcore.Field { return zap.String("method", s) }
+
+var (
+	defaultCoordinatorList = []int{7201, 7203, 7204}
+
+	defaultCoordinatorOptions = dockerResourceOptions{
+		source:        defaultCoordinatorSource,
+		containerName: defaultCoordinatorName,
+		dockerFile:    defaultCoordinatorDockerfile,
+		portList:      defaultCoordinatorList,
+	}
+)
+
+// Coordinator is a wrapper for a coordinator. It provides a wrapper on HTTP
+// endpoints that expose cluster management APIs as well as read and write
+// endpoints for series data.
+// TODO: consider having this work on underlying structures.
+type Coordinator interface {
+	Admin
+}
+
+// Admin is a wrapper for admin functions.
+type Admin interface {
+	// GetNamespace gets namespaces.
+	GetNamespace() (admin.NamespaceGetResponse, error)
+	// WaitForNamespace blocks until the given namespace is enabled.
+	// NB: if the name string is empty, this will instead
+	// check for a successful response.
+	WaitForNamespace(name string) error
+	// AddNamespace adds a namespace.
+	AddNamespace(admin.NamespaceAddRequest) (admin.NamespaceGetResponse, error)
+	// CreateDatabase creates a database.
+	CreateDatabase(admin.DatabaseCreateRequest) (admin.DatabaseCreateResponse, error)
+	// GetPlacement gets placements.
+	GetPlacement() (admin.PlacementGetResponse, error)
+	// WaitForPlacements blocks until the given placement IDs are present.
+	WaitForPlacements(ids []string) error
```

Comment:
Maybe we call this "WaitForInstances(..)"? Had to look at the implementation to get a sense of if it was instances in a placement or something else.